packages/story-editor/src/components/panels/layer/useLayers.js

```diff
@@ -14,30 +14,107 @@
  * limitations under the License.
  */
 
+/**
+ * External dependencies
+ */
+import { useMemo, useCallback } from '@googleforcreators/react';
+
 /**
  * Internal dependencies
  */
 import { useStory } from '../../../app';
-import { STABLE_ARRAY } from '../../../constants';
+import { nestedOffsetCalcFunc } from './constants';
 
 function useLayers() {
-  const elements = useStory(
-    ({ state }) =>
-      state.currentPage?.elements?.map(({ id, groupId, layerName }) => ({
+  const { elements, arrangeElement, arrangeGroup } = useStory(
+    ({ state, actions }) => ({
+      elements: state.currentPage?.elements || [],
+      arrangeElement: actions.arrangeElement,
+      arrangeGroup: actions.arrangeGroup,
+    })
+  );
+  const elementLayers = useMemo(
+    () =>
+      elements.map(({ id, groupId, layerName }, position) => ({
         id,
         groupId,
         layerName,
-      })) || STABLE_ARRAY
+        position,
+      })),
+    [elements]
+  );
+  elementLayers.reverse();
+
+  // This is a list of elements and groups in the order they're displayed
+  const layers = useMemo(() => {
+    const groupIds = new Set();
+    return (
+      elementLayers
+        // If an element has a new group, add both group and element
+        .map((element, index, list) => {
+          const elementLayer = {
+            isGroup: false,
+            isFirstElementAfterGroup:
+              index > 0 && Boolean(list[index - 1].groupId),
+            id: element.id,
+            position: element.position,
+          };
+          const isGroup = element.groupId && !groupIds.has(element.groupId);
+          if (!isGroup) {
+            return elementLayer;
+          }
+          groupIds.add(element.groupId);
+          const groupLayer = {
+            isGroup: true,
+            id: element.groupId,
+            position: element.position,
+            name: element.groupId,
+          };
+          return [groupLayer, elementLayer];
+        })
+        // Flatten the nested arrays
+        .flat()
+    );
+  }, [elementLayers]);
+
+  const handlePositionChange = useCallback(
```

Comment:
<!-- potential concern, ask for clarification or investigation -->
**❓question:**
I know it was brought over from another file and is unchanged, but does `handlePositionChange` here apply only when dragging a layer? Or could this be reapplied for keyboard events, too? Thinking that if it's just for dragging then would be good to specify the name for clarity (since we have #11906 for the keyboard handling).

Reply:
Renamed it - it is specifically for dragging only to handle the partial overlap and figure out what the new position is going to be.